src/Nonlinear/ReverseAD/Coloring/Coloring.jl

```diff
@@ -0,0 +1,547 @@
+module Coloring
+
+import DataStructures
+
+include("topological_sort.jl")
+
+# indexed sparse set of integers
+mutable struct IndexedSet
+    nzidx::Vector{Int}
+    empty::BitArray{1}
+    nnz::Int
+end
+
+IndexedSet(n::Integer) = IndexedSet(zeros(Int, n), trues(n), 0)
+
+function Base.push!(v::IndexedSet, i::Integer)
+    if v.empty[i]  # new index
+        v.nzidx[v.nnz+=1] = i
+        v.empty[i] = false
+    end
+    return
+end
+
+function Base.empty!(v::IndexedSet)
+    nzidx = v.nzidx
+    empty = v.empty
+    for i in 1:v.nnz
+        empty[nzidx[i]] = true
+    end
+    v.nnz = 0
+    return v
+end
+
+Base.length(v::IndexedSet) = length(v.nzidx)
+
+function Base.resize!(v::IndexedSet, n::Integer)
+    if n > length(v)
+        @assert v.nnz == 0 # only resize empty vector
+        resize!(v.nzidx, n)
+        resize!(v.empty, n)
+        fill!(v.empty, true)
+    end
+    return
+end
+
+Base.collect(v::IndexedSet) = v.nzidx[1:v.nnz]
+
+function Base.union!(v::IndexedSet, s)
+    for x in s
+        push!(v, x)
+    end
+    return
+end
+
+# compact storage for an undirected graph
+# neighbors of vertex i start at adjlist[offsets[i]]
+struct UndirectedGraph
+    adjlist::Vector{Int}
+    edgeindex::Vector{Int} # corresponding edge number, indexed as adjlist
+    offsets::Vector{Int}
+    edges::Vector{Tuple{Int,Int}}
+end
+
+_num_vertices(g::UndirectedGraph) = length(g.offsets) - 1
+
+_num_edges(g::UndirectedGraph) = length(g.edges)
+
+_num_neighbors(i::Int, g::UndirectedGraph) = g.offsets[i+1] - g.offsets[i]
+
+_start_neighbors(i::Int, g::UndirectedGraph) = g.offsets[i]
+
+function gen_adjlist(I, J, nel)
```

Comment:
since this is constructing an UndirectedGraph, it could be a constructor, or `graph_from_X(...`

Reply:
Changed to `UndirectedGraph` constructor.